File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js

//반대로 생각해도 된다.
function solution(triangle) {
    let leg = triangle[triangle.length-1].length;
    
    for(let i=leg-2;i>-1;i--)
    {
        for(let j=0;j<=i;j++)
        {
            triangle[i][j]+= Math.max(triangle[i+1][j],triangle[i+1][j+1]);   
        }
    }
    
    
    return triangle[0][0];
    
    
}






//dp
//테이블 설정
// arr[i][j] i,j번째 합의 최대값

//점화식
// dp[i][j] = Math.max( dp[i-1][j]+arr[i][j],dp[i-1][j-1]+arr[i][j])

//초기값
//arr[0][0]=7;

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    
    static int[][] dp;
    public int solution(int[][] triangle) {
        int answer=0;
        int r= triangle.length;
        int c= triangle[triangle.length-1].length;
        dp = new int[r][c];
        
        dp[0][0]=triangle[0][0];
        for(int i=1;i<triangle.length;i++)
        {
            dp[i][0]=dp[i-1][0] + triangle[i][0];
            for(int j=1;j<=i;j++)
            {
                dp[i][j]= Math.max(dp[i-1][j] +  triangle[i][j], dp[i-1][j-1] +  triangle[i][j]);
            }
        }
        
        for(int i=0;i<c;i++)
        {
            answer = Math.max(dp[r-1][i],answer);
        }
        
        return answer;
    }
}

//테이블 설정
// arr[i][j] i,j에 오는 삼각형의 합의 최대값

//dp[i][j]=  max (dp[i-1][j] + arr[i][j], dp[i-1][j-1] + arr[i][j])